src/preprocessing/shapes/triangle_mesh.jl

```diff
@@ -0,0 +1,146 @@
+# This is the data format returned by `load(file)` when used with `.stl` files
+struct TriangleMesh{NDIMS, ELTYPE}
+    vertices          :: Vector{SVector{NDIMS, ELTYPE}}
+    face_vertices     :: Vector{NTuple{3, SVector{NDIMS, ELTYPE}}}
+    face_vertices_ids :: Vector{NTuple{3, Int}}
+    face_edges_ids    :: Vector{NTuple{3, Int}}
+    edge_vertices_ids :: Vector{NTuple{2, Int}}
+    vertex_normals    :: Vector{SVector{NDIMS, ELTYPE}}
+    edge_normals      :: Vector{SVector{NDIMS, ELTYPE}}
+    face_normals      :: Vector{SVector{NDIMS, ELTYPE}}
+    min_corner        :: SVector{NDIMS, ELTYPE}
+    max_corner        :: SVector{NDIMS, ELTYPE}
+
+    function TriangleMesh(face_vertices, face_normals, vertices)
+        NDIMS = 3
+
+        return TriangleMesh{NDIMS}(face_vertices, face_normals, vertices)
+    end
+
+    # Function barrier to make `NDIMS` static and therefore `SVector`s type-stable
+    function TriangleMesh{NDIMS}(face_vertices, face_normals, vertices) where {NDIMS}
+        ELTYPE = eltype(first(face_normals))
+        n_faces = length(face_normals)
+
+        face_vertices_ids = fill((0, 0, 0), n_faces)
+
+        @threaded face_vertices for i in 1:n_faces
+            v1 = face_vertices[i][1]
+            v2 = face_vertices[i][2]
+            v3 = face_vertices[i][3]
+
+            # TODO: This part is about 90% of the runtime.
+            # Vertex IDs are only needed for the hierarchical winding.
+            vertex_id1 = findfirst(x -> v1 == x, vertices)
+            vertex_id2 = findfirst(x -> v2 == x, vertices)
+            vertex_id3 = findfirst(x -> v3 == x, vertices)
```

Comment:
```suggestion
            # We removed duplicate vertices with an absolute tolerance of 1e-14.
            # `searchsortedfirst` finds the first value in `vertices` greater than or equivalent to the first argument.
            # We might have removed `v1` because there was `v1 - 1e-14` in `vertices`, so we can find
            # this value by subtracting `1e-14`.
            vertex_id1 = searchsortedfirst(v1 - 1e-14, vertices)
            vertex_id2 = searchsortedfirst(v2 - 1e-14, vertices)
            vertex_id3 = searchsortedfirst(v3 - 1e-14, vertices)
```